### PR TITLE
Gradient clipping

### LINKED ISSUE #6: Gradient clipping support
BaseModel._create_train_ops should be configurable to perform gradient clipping. This will allow us to implement and train more advanced models such as seq2seq etc.

### PR DESCRIPTION
**Pull request by bedapisl**
Closes #6.

### CODE REVIEW

File: emloop_tensorflow/model.py [hunk 1]
```diff
@@ -375,8 +375,12 @@ def _create_train_ops(self, dependencies: List[List[tf.Operation]], optimizer_co
         if optimizer_config is None:
             raise ValueError('Optimizer config was not specified although it is required for creating the train op. '
                              'Please specify the configuration in `model.optimizer`.')
+
+        optimizer_config_copy = optimizer_config.copy()
+        clip = optimizer_config_copy.pop("clip", None)
+
         grads_and_vars = []
-        optimizer = create_optimizer(optimizer_config)
+        optimizer = create_optimizer(optimizer_config_copy)
```
Comment (FloopCZ) [inline emloop_tensorflow/model.py]: Why is it necessary to pass a copy of the config here? Wouldn't it be sufficient to just extract the `clip` via `optimizer_config.get('clip', None)` ?

File: emloop_tensorflow/model.py [hunk 2]
```diff
@@ -385,6 +389,14 @@ def _create_train_ops(self, dependencies: List[List[tf.Operation]], optimizer_co
         for tower in self._towers:
             with tower:
                 grads_and_vars.append(optimizer.compute_gradients(tf.reduce_mean(tower.loss) + regularization_loss))
+
+        # gradient clipping
+        if clip is not None:
+            for i, tower_grads_vars in enumerate(grads_and_vars):
```
Comment (FloopCZ) [inline emloop_tensorflow/model.py]:
```suggestion
            for tower_grads_vars in grads_and_vars:
```

File: emloop_tensorflow/model.py [hunk 3]
```diff
@@ -385,6 +389,14 @@ def _create_train_ops(self, dependencies: List[List[tf.Operation]], optimizer_co
...
+                for j, (grad, var) in enumerate(tower_grads_vars):
+                    if grad is not None:
+                        grads_and_vars[i][j] = (tf.clip_by_value(grad, -clip, clip), var)
```
Comment (FloopCZ) [inline emloop_tensorflow/model.py]:
```suggestion
                        tower_grad_vars[i] = (tf.clip_by_value(grad, -clip, clip), var)
```

File: emloop_tensorflow/model.py [hunk 4]
```diff
@@ -385,6 +389,14 @@ def _create_train_ops(self, dependencies: List[List[tf.Operation]], optimizer_co
         for tower in self._towers:
             with tower:
                 grads_and_vars.append(optimizer.compute_gradients(tf.reduce_mean(tower.loss) + regularization_loss))
+
+        # gradient clipping
+        if clip is not None:
+            for i, tower_grads_vars in enumerate(grads_and_vars):
+                for j, (grad, var) in enumerate(tower_grads_vars):
```
Comment (FloopCZ) [inline emloop_tensorflow/model.py]:
```suggestion
                for i, (grad, var) in enumerate(tower_grads_vars):
```

File: emloop_tensorflow/model.py (hunk 1 above)
Comment (blazekadam) [reply emloop_tensorflow/model.py]: Well, the error @bedapisl mentioned is quite natural since the `optimizer_config` is passed as-is to the selected `tf` optimizer constructor.

It would prefer a documented `Model` argument (`clip_gradient`?) so that we can avoid this...

File: emloop_tensorflow/model.py [hunk 5]
```diff
@@ -81,6 +81,7 @@ def __init__(self,  # pylint: disable=too-many-arguments
         :param loss_name: loss tensor name
         :param monitor: monitor signal mean and variance of the tensors which names contain the specified value
         :param restore_fallback: ignored arg. (allows training from configs saved by emloop where it is added)
+        :param clip_gradient: How much gradients are clipped, None for no clipping
```
Comment (FloopCZ) [inline emloop_tensorflow/model.py]:
```suggestion
        :param clip_gradient: limit the absolute value of the gradient; set to None for no clipping
```

File: emloop_tensorflow/model.py [hunk 6]
```diff
@@ -375,6 +377,7 @@ def _create_train_ops(self, dependencies: List[List[tf.Operation]], optimizer_co
         if optimizer_config is None:
             raise ValueError('Optimizer config was not specified although it is required for creating the train op. '
                              'Please specify the configuration in `model.optimizer`.')
+
```
Comment (blazekadam) [inline emloop_tensorflow/model.py]:
```suggestion
```

File: emloop_tensorflow/model.py [hunk 7]
```diff
@@ -81,6 +81,7 @@ def __init__(self,  # pylint: disable=too-many-arguments
         :param loss_name: loss tensor name
         :param monitor: monitor signal mean and variance of the tensors which names contain the specified value
         :param restore_fallback: ignored arg. (allows training from configs saved by emloop where it is added)
+        :param clip_gradient: limit the absolute value of the gradient; set to None for no clipping
```
Comment (blazekadam) [inline emloop_tensorflow/model.py]:
```suggestion
        :param clip_gradient: limit the absolute value of gradients to the given value; set to None for no clipping
```
